viewer/src/views/threejs/cad/CadNode.ts

```diff
@@ -44,11 +45,17 @@ export class CadNode extends THREE.Object3D {
   private readonly _previousCameraMatrix = new THREE.Matrix4();
   private readonly _boundingBoxNode: THREE.Object3D;
 
+  // TODO make private + friend?
+  public readonly _materials: Materials;
```

Comment:
This should be fixed before merge IMO.

Reply:
Do you have any suggestions on how to fix it? I am a bit stuck with my own mindset here and cannot seem to figure out how to redesign it. Friends are not a thing in typescript (and often a bad thing to use anyways).

Reply:
What about adding 1) get/set CadNode.RenderType as a field which traverses through the materials or 2) add a RenderType-field to the RenderingHints-struct which does the same as 1?

I think I like 1) more because RenderType (which maybe should be renamed to RenderMode?) because it's not really a hint, but a requirement.

Reply:
But overriding the RenderType/Mode is really not something that should be done by an application developer - especially not if an effect is also trying to override the material. This is where it would be so much better to inject the material into the render function, but we need to do the whole traversal to change material + traversal to revert the change because of the way Three.js is designed.

Given that we have to do the forward change and back again, I really do not want the user to be able to change the material and have our "internal" effects change them too.

That being said, ideally we should not have had to change the material at all, since it will never be rendered "normally" anymore :confused: I do not know if we can achieve that, though.

Reply:
And yes, 1) is better than 2) since 2) makes changing the material to the packed version seem like a common thing to do, which I do not think it will be.

Reply:
1) done